File: cli/src/commands/watch.ts
/**
 * mrsf watch — Monitor files and validate / reanchor on change.
 *
 * By default, validation is run on every change.  Pass `--reanchor` to
 * also auto-fix drifted anchors when a Markdown file is saved.
 *
 * Trigger rules (avoids feedback loops):
 *   - `.md` change       → reanchor (if enabled), then validate sidecar
 *   - `.review.yaml/json` change → validate only
 */

import type { Command } from "commander";
import path from "node:path";
import chalk from "chalk";
import { watch as chokidarWatch, type FSWatcher } from "chokidar";

import {
  findWorkspaceRoot,
  discoverSidecar,
  sidecarToDocument,
  discoverAllSidecars,
} from "../lib/discovery.js";
import { resolveSidecarPaths } from "../lib/resolve-files.js";
import { validateFile } from "../lib/validator.js";
import { reanchorFile } from "../lib/reanchor.js";
import type { ReanchorResult, ValidationResult } from "../lib/types.js";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

const SIDECAR_EXTS = [".review.yaml", ".review.json"];
function isSidecar(file: string): boolean {
  return SIDECAR_EXTS.some((ext) => file.endsWith(ext));
}

function timestamp(): string {
  return new Date().toLocaleTimeString("en-GB", { hour12: false });
}

function statusIcon(status: string): string {
  switch (status) {
    case "anchored":
      return chalk.green("●");
    case "shifted":
      return chalk.blue("↕");
    case "fuzzy":
      return chalk.yellow("≈");
    case "ambiguous":
      return chalk.magenta("?");
    case "orphaned":
      return chalk.red("✗");
    default:
      return "·";
  }
}

function printReanchorResults(
  label: string,
  results: ReanchorResult[],
  changed: number,
  written: boolean,
  dryRun: boolean,
  verbose: boolean,
): void {
  const prefix = dryRun ? chalk.dim("[dry-run] ") : "";
  console.log(`  ${prefix}${chalk.cyan("reanchor")} ${label}`);
  for (const r of results) {
    const icon = statusIcon(r.status);
    const line = r.newLine != null ? `:${r.newLine}` : "";
    console.log(`    ${icon} ${r.commentId}${line} — ${r.reason}`);
    if (verbose && r.score != null) {
      console.log(`        score: ${r.score.toFixed(3)}`);
    }
  }
  if (!dryRun && written) {
    console.log(`  ${chalk.green("✓")} ${changed} comment(s) updated`);
  } else if (dryRun) {
    console.log(`  ${chalk.dim(`${changed} comment(s) would change`)}`);
  }
}

function printValidateResults(
  label: string,
  result: ValidationResult,
  strict: boolean,
): void {
  const effectiveValid = strict
    ? result.valid && result.warnings.length === 0
    : result.valid;
  const icon = effectiveValid ? chalk.green("✓") : chalk.red("✗");
  console.log(`  ${chalk.cyan("validate")} ${icon} ${label}`);

  for (const e of result.errors) {
    console.log(
      chalk.red(`    error: ${e.message}${e.path ? ` (${e.path})` : ""}`),
    );
  }
  for (const w of result.warnings) {
    const fn = strict ? chalk.red : chalk.yellow;
    const tag = strict ? "error" : "warn";
    console.log(
      fn(`    ${tag}: ${w.message}${w.path ? ` (${w.path})` : ""}`),
    );
  }
}

// ---------------------------------------------------------------------------
// Command
// ---------------------------------------------------------------------------

export interface WatchOptions {
  reanchor?: boolean;
  dryRun?: boolean;
  threshold: string;
  git: boolean;
  from?: string;
  updateText?: boolean;
  force?: boolean;
  strict?: boolean;
  debounce: string;
}

export function registerWatch(program: Command): void {
  program
    .command("watch [files...]")
    .description(
      "Watch files for changes and validate (+ optionally reanchor)",
    )
    .option("--reanchor", "Auto-reanchor when Markdown files change")
    .option("-n, --dry-run", "Preview reanchor changes without writing")
    .option(
      "-t, --threshold <n>",
      "Fuzzy match threshold 0.0–1.0",
      "0.6",
    )
    .option("-s, --strict", "Treat validation warnings as errors")
    .option("--no-git", "Disable git integration")
    .option("--from <commit>", "Override from-commit for reanchor")
    .option("--update-text", "Update selected_text on reanchor")
    .option(
      "-f, --force",
      "Force reanchor — update commit, clear audit fields",
    )
    .option("--debounce <ms>", "Debounce interval in ms", "300")
    .action(async (files: string[], opts: WatchOptions) => {
      const parentOpts = program.opts();
      const cwd = parentOpts.cwd ?? process.cwd();
      const quiet = parentOpts.quiet ?? false;
      const verbose = parentOpts.verbose ?? false;
      const debounceMs = parseInt(opts.debounce, 10) || 300;
      const threshold = parseFloat(opts.threshold);
      const noGit = !opts.git;

      // ── Resolve watch targets ────────────────────────────
      let watchPaths: string[];

      if (files.length === 0) {
        // Watch the whole workspace
        const root = findWorkspaceRoot(cwd) ?? cwd;
        watchPaths = [root];
      } else {
        // Resolve explicit paths — collect both sidecars and documents
        const sidecarPaths = await resolveSidecarPaths(files, cwd);
        const docPaths = sidecarPaths.map((sp) => sidecarToDocument(sp));
        watchPaths = [...new Set([...sidecarPaths, ...docPaths])];
      }

      // ── Start watcher ────────────────────────────────────
      const watcher = chokidarWatch(watchPaths, {
        ignoreInitial: true,
        ignored: [
          "**/node_modules/**",
          "**/.git/**",
          ...(files.length === 0
            ? []
            : [
                // When watching specific files, chokidar might get
                // extra events from parent dirs — no extra ignores needed.
              ]),
        ],
        // Only watch files matching our extensions when watching a dir
        ...(files.length === 0
          ? {}
          : {}),
      });

      // Track debounce per file
      const timers = new Map<string, ReturnType<typeof setTimeout>>();
      // Track files we just wrote (to skip re-trigger)
      const selfWrites = new Set<string>();
      // Stats
      let totalEvents = 0;
      let totalErrors = 0;
      let totalReanchors = 0;

      // ── Event handler ────────────────────────────────────
      async function handleChange(filePath: string): Promise<void> {
        const abs = path.resolve(filePath);
        const rel = path.relative(cwd, abs);

        // Skip files we just wrote ourselves
        if (selfWrites.has(abs)) {
          selfWrites.delete(abs);
          return;
        }

        // Only process markdown and sidecar files
        const isSidecarFile = isSidecar(abs);
        const isMarkdown = abs.endsWith(".md");
        if (!isSidecarFile && !isMarkdown) return;

        totalEvents++;
        if (!quiet) {
          console.log(
            `\n${chalk.dim(`[${timestamp()}]`)} ${chalk.bold(rel)}`,
          );
        }

        if (isSidecarFile) {
          // ── Sidecar changed → validate ──────────────────
          await runValidate(abs, rel);
        } else if (isMarkdown) {
          // ── Markdown changed → reanchor (opt-in) + validate sidecar ──
          let sidecarPath: string | undefined;
          try {
            sidecarPath = await discoverSidecar(abs, { cwd });
          } catch {
            // No sidecar for this markdown file — skip silently
            if (verbose) {
              console.log(chalk.dim(`  no sidecar found — skipping`));
            }
            return;
          }

          if (opts.reanchor) {
            await runReanchor(sidecarPath, rel);
          }

          // Validate the sidecar in all modes
          await runValidate(
            sidecarPath,
            path.relative(cwd, sidecarPath),
          );
        }
      }

      async function runValidate(
        sidecarPath: string,
        label: string,
      ): Promise<void> {
        try {
          const result = await validateFile(sidecarPath);
          const effectiveValid = opts.strict
            ? result.valid && result.warnings.length === 0
            : result.valid;
          if (!effectiveValid) totalErrors++;

          if (!quiet || !effectiveValid) {
            printValidateResults(label, result, !!opts.strict);
          }
        } catch (err: unknown) {
          totalErrors++;
          const msg = err instanceof Error ? err.message : String(err);
          console.error(chalk.red(`  validate ✗ ${label}: ${msg}`));
        }
      }

      async function runReanchor(
        sidecarPath: string,
        label: string,
      ): Promise<void> {
        try {
          const { results, changed, written } = await reanchorFile(
            sidecarPath,
            {
              cwd,
              dryRun: opts.dryRun,
              threshold,
              noGit,
              fromCommit: opts.from,
              updateText: opts.updateText,
              force: opts.force,
            },
          );

          totalReanchors++;

          if (written) {
            // Mark the sidecar as self-written to avoid re-triggering
            selfWrites.add(path.resolve(sidecarPath));
          }

          const orphaned = results.filter(
            (r) => r.status === "orphaned",
          ).length;
          if (orphaned > 0) totalErrors++;

          if (!quiet) {
            printReanchorResults(
              label,
              results,
              changed,
              written,
              !!opts.dryRun,
              verbose,
            );
          }
        } catch (err: unknown) {
          totalErrors++;
          const msg = err instanceof Error ? err.message : String(err);
          console.error(chalk.red(`  reanchor ✗ ${label}: ${msg}`));
        }
      }

      // ── Wire events with debounce ────────────────────────
      function onFileChange(filePath: string): void {
        const abs = path.resolve(filePath);
        const existing = timers.get(abs);
        if (existing) clearTimeout(existing);

        timers.set(
          abs,
          setTimeout(() => {
            timers.delete(abs);
            handleChange(abs).catch((err) => {
              console.error(
                chalk.red(`  unexpected error: ${String(err)}`),
              );
            });
          }, debounceMs),
        );
      }

      watcher.on("change", onFileChange);
      watcher.on("add", onFileChange);

      // ── Startup banner ───────────────────────────────────
      if (!quiet) {
        const mode = opts.reanchor
          ? opts.dryRun
            ? "validate + reanchor (dry-run)"
            : "validate + reanchor"
          : "validate";
        const pathDesc =
          files.length === 0
            ? "workspace"
            : `${watchPaths.length} path(s)`;
        console.log(
          chalk.green(`Watching ${pathDesc} — mode: ${chalk.bold(mode)}`),
        );
        console.log(chalk.dim("Press Ctrl+C to stop.\n"));
      }

      // ── Graceful shutdown ────────────────────────────────
      async function shutdown(): Promise<void> {
        // Clear pending debounce timers
        for (const t of timers.values()) clearTimeout(t);
        timers.clear();

        await watcher.close();

        if (!quiet) {
          console.log(
            `\n${chalk.dim("──────────────────────────────────────")}`,
          );
          console.log(
            `Events processed: ${totalEvents}  |  ` +
              `Reanchors: ${totalReanchors}  |  ` +
              `Errors: ${totalErrors}`,
          );
        }

        process.exit(totalErrors > 0 ? 1 : 0);
      }

      process.on("SIGINT", () => void shutdown());
      process.on("SIGTERM", () => void shutdown());

      // Keep the process alive
      await new Promise<void>(() => {
        // Never resolves — the process runs until interrupted
      });
    });
}

File: python/src/mrsf/cli/watch.py
"""mrsf watch [files...] — watch for changes and validate/reanchor."""

from __future__ import annotations

import os
import sys
import time
from datetime import datetime
from typing import Any

import click

from ..discovery import discover_sidecar
from ..reanchor import reanchor_file
from ..resolve_files import resolve_sidecar_paths
from ..types import ReanchorOptions
from ..validator import validate_file
from .main import MrsfContext, pass_ctx
from .watch_signals import register_signal_handlers


@click.command("watch")
@click.argument("files", nargs=-1, type=click.Path())
@click.option("--reanchor", "do_reanchor", is_flag=True, help="Reanchor on markdown changes.")
@click.option("-n", "--dry-run", is_flag=True, help="Report without modifying files.")
@click.option("-t", "--threshold", type=float, default=0.6, help="Fuzzy match threshold.")
@click.option("-s", "--strict", is_flag=True, help="Treat warnings as errors.")
@click.option("--no-git", is_flag=True, help="Disable git integration.")
@click.option("--from", "from_commit", default=None, help="Base commit for diff.")
@click.option("--update-text", is_flag=True, help="Update selected_text with current text.")
@click.option("-f", "--force", is_flag=True, help="Force-anchor high-confidence results.")
@click.option("--debounce", type=int, default=300, help="Debounce interval in milliseconds.")
@pass_ctx
def watch_cmd(
    ctx: MrsfContext,
    files: tuple[str, ...],
    do_reanchor: bool,
    dry_run: bool,
    threshold: float,
    strict: bool,
    no_git: bool,
    from_commit: str | None,
    update_text: bool,
    force: bool,
    debounce: int,
) -> None:
    """Watch for changes and validate/reanchor sidecars."""
    try:
        from watchdog.events import FileSystemEventHandler
        from watchdog.observers import Observer
    except ImportError:
        click.echo(
            click.style(
                "watchdog is required for watch command. Install with: pip install watchdog",
                fg="red",
            ),
            err=True,
        )
        sys.exit(1)

    sidecars = resolve_sidecar_paths(list(files), cwd=ctx.cwd)
    # Also watch the markdown files
    watch_paths = set()
    for sc in sidecars:
        watch_paths.add(os.path.dirname(sc) or ".")

    if not watch_paths:
        watch_paths.add(ctx.cwd)

    # Track self-writes to avoid re-triggering
    self_writes: dict[str, float] = {}
    debounce_timers: dict[str, float] = {}
    event_count = 0
    error_count = 0
    reanchor_count = 0

    def handle_file(file_path: str) -> None:
        nonlocal event_count, error_count, reanchor_count

        abs_path = os.path.abspath(file_path)

        # Skip self-writes (within a 2-second window to handle multiple FS events)
        sw_time = self_writes.get(abs_path)
        if sw_time is not None and (time.time() - sw_time) < 2.0:
            return

        # Debounce
        now = time.time()
        last = debounce_timers.get(abs_path, 0)
        if now - last < debounce / 1000.0:
            return
        debounce_timers[abs_path] = now

        event_count += 1
        ts = datetime.now().strftime("%H:%M:%S")

        is_sidecar = abs_path.endswith(".review.yaml") or abs_path.endswith(".review.json")
        is_markdown = abs_path.endswith(".md") or abs_path.endswith(".markdown")

        if is_sidecar:
            if not ctx.quiet:
                click.echo(f"[{ts}] validate {os.path.basename(abs_path)}")
            try:
                result = validate_file(abs_path)
            except Exception as e:
                # File may be mid-write; skip silently
                if ctx.verbose:
                    click.echo(click.style(f"  validate skipped (transient): {e}", dim=True))
                return
            if not result.valid or (strict and result.warnings):
                error_count += 1
                for err in result.errors:
                    path_suffix = f" ({err.path})" if err.path else ""
                    click.echo(click.style(f"  ERROR: {err.message}{path_suffix}", fg="red"))

        elif is_markdown:
            # Find companion sidecar
            try:
                sidecar = discover_sidecar(abs_path, cwd=ctx.cwd, config_path=ctx.config)
            except Exception:
                return

            if not os.path.exists(sidecar):
                return

            if do_reanchor:
                if not ctx.quiet:
                    prefix = "[dry-run] " if dry_run else ""
                    click.echo(f"[{ts}] {prefix}reanchor {os.path.basename(abs_path)}")
                try:
                    opts = ReanchorOptions(
                        cwd=ctx.cwd,
                        dry_run=dry_run,
                        threshold=threshold,
                        no_git=no_git,
                        from_commit=from_commit,
                        update_text=update_text,
                        force=force,
                    )
                    results, changed, written = reanchor_file(sidecar, opts)
                    reanchor_count += 1
                    if written:
                        self_writes[os.path.abspath(sidecar)] = time.time()
                except Exception as e:
                    error_count += 1
                    click.echo(click.style(f"  Error: {e}", fg="red"))

            # Validate sidecar too
            if not ctx.quiet:
                click.echo(f"[{ts}] validate {os.path.basename(sidecar)}")
            try:
                result = validate_file(sidecar)
            except Exception as e:
                if ctx.verbose:
                    click.echo(click.style(f"  validate skipped (transient): {e}", dim=True))
                return
            if not result.valid or (strict and result.warnings):
                error_count += 1

    class Handler(FileSystemEventHandler):
        def on_modified(self, event: Any) -> None:
            if not event.is_directory:
                handle_file(event.src_path)

        def on_created(self, event: Any) -> None:
            if not event.is_directory:
                handle_file(event.src_path)

    observer = Observer()
    handler = Handler()
    for wp in watch_paths:
        observer.schedule(handler, wp, recursive=True)

    observer.start()
    if not ctx.quiet:
        click.echo(f"Watching {len(watch_paths)} path(s)... Press Ctrl+C to stop.")

    def shutdown(*_: Any) -> None:
        cleanup_signal_handlers()
        observer.stop()
        if not ctx.quiet:
            click.echo(
                f"\n{event_count} event(s), {reanchor_count} reanchor(s), {error_count} error(s)"
            )
        sys.exit(1 if error_count > 0 else 0)

    cleanup_signal_handlers = register_signal_handlers(shutdown)

    try:
        observer.join()
    except KeyboardInterrupt:
        shutdown()
    finally:
        cleanup_signal_handlers()
